juju_spell/commands/grant.py

```
@@ -14,21 +14,60 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 """Command to grant permission for user."""
-from typing import Any
+from typing import Any, List, Optional
 
 from juju.controller import Controller
 
 from juju_spell.commands.base import BaseJujuCommand
 
 __all__ = ["GrantCommand", "ACL_CHOICES"]
 
-ACL_CHOICES = ["login", "add-model", "superuser"]
+CONTROLLER_ACL_CHOICES = ["login", "add-model", "superuser"]
+MODEL_ACL_CHOICES = ["read", "write", "admin"]
+
+ACL_CHOICES = CONTROLLER_ACL_CHOICES + MODEL_ACL_CHOICES
 
 
 class GrantCommand(BaseJujuCommand):
     """Grant permission for user."""
 
-    async def execute(self, controller: Controller, *args: Any, **kwargs: Any) -> bool:
+    async def execute(
+        self,
+        controller: Controller,
+        *args: Any,
+        models: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> bool:
         """Execute."""
-        result: bool = await controller.grant(username=kwargs["user"], acl=kwargs["acl"])
-        return result
+        acl = kwargs["acl"]
+        controller_acl: str
+        model_acl: str
+        if acl in CONTROLLER_ACL_CHOICES:
+            controller_acl = acl
+        else:
+            controller_acl = "login"
```

We should not silently change the acl value here, since the validation is done in CLI it should never be anything else and if yes than it should raise an error.

---

If we are trying to do some default values, than it should be like `if acl is None` or even better is `kwargs.get("acl", "login")`

---

> We should not silently change the acl value here, since the validation is done in CLI it should never be anything else and if yes than it should raise an error.

Here is the command, so we are not change the value.

Because we are grant user's acl to model and controller in the same time.

The acl value can be either model permission in [`read`, `write`, `admin`] or controller permission in [`superuser`, `login`]. The idea is user only need to type one of the acl then we will generate the correspond acl for user in both model and controller.

---

We should at-least moved this logic to separate function or make it simpler, because like this we have too many if statements.

---

Updated